### DISCUSSION TITLE
Unable to set request body optionally

### DISCUSSION THREAD

**Discussion by bensullivan-ag** [Kong/unirest-java#529] [Q&A]
Hi

Should the following pass?

```
@Test
void shouldAllowRequestBodiesToBeAppliedConditionally() {
    var maybeBody = Optional.of("test body");
    var request = Unirest.patch("api.com").header("Authorization", "Bearer XXXXXX");
    maybeBody.ifPresent((b) -> request.header("Content-Type", "application/json").body(b));
    assertTrue(request.getBody().isPresent());
    assertEquals(maybeBody.get(), request.getBody().get());
}
```

Thanks

**Comment by ryber** [comment]
It would if you configured an Object Mapper to understand what an Optional
is. Otherwise it’s just going to call toString on it.

**Comment by bensullivan-ag** [comment]
Thanks for reply.

Which line are you referring to?

`b` is a `String` and `request.getBody()` is an `Optional`..

**Comment by ryber** [comment]
Sorry I was reading it from my phone, I thought you were passing the Optional as the body. You don't even need the Optional at all to produce the same result. It could be this:

```
@Test
    void shouldAllowRequestBodiesToBeAppliedConditionally() {
        var request = Unirest.patch(MockServer.PATCH).header("Authorization", "Bearer XXXXXX");
        request.header("Content-Type", "application/json").body("test body");
        assertTrue(request.getBody().isPresent());
        assertEquals("test body", request.getBody().get());
    }
```
but the answer is no, this would not work. "Test body" is a string, ```request.getBody().get()``` is a Body object. In this case you are looking at what Unirest is planning to ship. What "Body" actually represents is pretty complicated because it could include multipart forms. Additionally at the level you are working it will probably produce a empty optional as you are in the middle of the request process.

**Comment by ryber** [comment]
Oh another thing, the Unirest request object goes though different transformations as you build it. What is returned from .body("") is a DIFFERENT object from your request object. So the original "request" variable won't have a body, because it is a previous version of the chain.
